CSI: remove global vars

hoegaarden opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This commit removes global state in the variables
- `csi.csiDrivers`
- `csi.nim`
- `csi.PluginHandler`

The CSI plugin struct now holds the methods to implement a volume plugin
and a plugin handler. The state formerly held in the package
global variables now has been moved into this struct, too.
Therefore, any instance of the CSI plugin now is completely indipendent
and does not share anything with any other instance.

Because the CSI plugin is a very special plugin, the volume plugin
manager knows some specifics about is: it has a method to specifically
get hold of the plugin and return its plugin handler aspect.
This is used by the kubelet to register this plugin handler with its
plugin watcher without the need to know specifics about the CSI plugin,
all those are handled in the volume plugin manager.

Additional notes:

- Because the the kubelet fails to start now if the CSI
  plugin cannot be found, a CSI plugin now also needs to be registered
  for kubemark's hollow kubelet. We use a faked plugin here.
- The fake implementations (for the plugin manager & kubelet watchable
  volume plugin) have been automatically generated by `counterfeiter` and
  can therefore be updated automatically.
- To break an import cycle (the CSI plugin imports the volume plugin,
  now that the volume plugin manager needs to query for the CSI plugin,
  it in turn imports the CSI plugin -> import cycle) we've split out the
  const for the plugin name into its own package. This pacakge can be
  imported by all packages in question: the volume & the csi package.

<!-- **Special notes for your reviewer** -->


**Release notes**

```release-note
NONE
```

/assign @vladimirvivien

hoegaarden committed:
csi: Remove package global vars

This commit removes global state in the variables
- `csi.csiDrivers`
- `csi.nim`
- `csi.PluginHandler`

The CSI plugin struct now holds the methods to implement a volume plugin
and a plugin handler. The state formerly held in the package
global variables now has been moved into this struct, too.
Therefore, any instance of the CSI plugin now is completely indipendent
and does not share anything with any other instance.

Because the CSI plugin is a very special plugin, the volume plugin
manager knows some specifics about it: it has a method to specifically
get hold of the plugin and return its plugin handler aspect.
This is used by the kubelet to register this plugin handler with its
plugin watcher without the need to know specifics about the CSI plugin,
all those are handled in the volume plugin manager.

Additional notes:

- If the plugin handler cannot be registered at the start of the
  kubelet we retry the registration every 10 secs in the background.
  For each failed registration attempt we generate a log message and an
  event
- Because the the kubelet fails to start now if the CSI
  plugin cannot be found, a CSI plugin now also needs to be registered
  for kubemark's hollow kubelet. We use a faked plugin here.
- The fake implementations (for the plugin manager & kubelete watchable
  volume plugin) have been automatically generated by `counterfeiter` and
  can therefore be updated automatically.
- To break an import cycle (the CSI plugin imports the volume plugin,
  now that the volume plugin manager needs to query for the CSI plugin,
  it in turn imports the CSI plugin -> import cycle) we've split out the
  const for the plugin name into its own package. This pacakge can be
  imported by all packages in question: the volume & the csi package.

hoegaarden commented:
/test pull-kubernetes-integration

hoegaarden commented:
/milestone v1.15

hoegaarden commented:
/test pull-kubernetes-e2e-gce
boskos seems to have issues

hoegaarden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hoegaarden commented:
/retest

hoegaarden commented:
/retest

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -197,15 +202,62 @@ func (h *RegistrationHandler) validateVersions(callerName, pluginName string, en
+
+// getClientCreator takes a driver name and returns a function that constructs
+// a new CSI driver client of that name/type.
+func (p *csiPlugin) getClientCreator(driverName csiDriverName) func() (*csiDriverClient, error) {
```
Do we still need this level of indirection when creating the client.  Why not use `csi_plugin.newCsiDriverClient` directly?

hoegaarden replied:
This func is there, to help a bit with the `csiClientGetter`, e.g. like here: https://github.com/kubernetes/kubernetes/pull/74966/files#diff-f52821ec6236e86b02ca334533395c91R523.

Would you prefer to store both
- the driver name
- the `newCsiDriverClient` func
on the `csiClientGetter` struct and call the latter with the former on a `csiClientGetter.Get()` call?

Happy to do so. TBH it feels a bit nicer to me as it is, but I have no strong opinions about that.

hoegaarden replied:
We can see if we can streamline the lazy creation and/or caching of the client in a separate PR later, we might be able to combine different aspects of `csiPlugin`, `csiClientGetter`, `getClientCreator`, and `newCsiDriverClient` -- and make the code easier to understand by doing so.

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -907,3 +1014,15 @@ func isV1Version(version string) bool {
+
+// ToPluginHandler converts the generic volume plugin to a specific csiPlugin and
+// returns that as a PluginHandler. In case the generic plugin is not a CSI
+// plugin and thus not a PluginHandler, an error is returned.
```
Please indicate in comment that this is used by pluginwatcher as a way to get access to the csiPlugin.

vladimirvivien commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -1514,3 +1522,55 @@ func TestHighestSupportedVersion(t *testing.T) {
+	}{
+		"nil cannot be type-casted to csiPlugin": {
+			genericPlugin:  nil,
+			expectedErrMsg: "is not a CSI volume plugin",
```
Would test be easier if you simply check for existence of err instead of comparing error string ? A simple string change in error string (case change, error correction, etc) will cause test to fail.

hoegaarden replied:
Yeah, I had different error cases in mind somewhen. But that's not the case anyway, so absolutely, we should just check on error or not here.

vladimirvivien reviewed (commented):
Ok let's start here.  The code looks good, but I left a few comments.

hoegaarden commented on pkg/controller/.import-restrictions:
```diff
@@ -101,13 +101,18 @@
         "github.com/google/gofuzz",
         "github.com/prometheus/client_golang/prometheus",
         "github.com/robfig/cron",
+        "github.com/spf13/afero",
```
No -- it has nothing to do with the kubelet change. The reasons is:

For the `ToPlunginHandler` function, we need the type `pluginwatcher.PluginHandler`, therefore we import the package `pluginwatcher` in the `csi` package. A bit more about that dependencies can be found [in the other comment](https://github.com/kubernetes/kubernetes/pull/73919#discussion_r269077222), but in short: those new dependencies are all transitive dependencies, brought in by the `pluginwatcher` package and because the `controller/volume/...` package has a dependency on the `csi` package.

hoegaarden commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -77,36 +81,35 @@ type driverMode string
+// PluginHandler is only needed for the kubelet to add the CSI plugin as a pluginwatcher handler
+// as soon as the kubelet discovers the CSI plugin via name and not via this
+// package-global variable, we can remove that singleton again.
+// TODO(hoegaarden) remove when kubelet discovers the CSI plugin handler by name
```
This capacity is added by https://github.com/kubernetes/kubernetes/pull/74966/commits/94859e7e00456391030c067b011c7922b453a08a and the "workaround" (`sync.Once` & global `PluginHandler`) and therefore also the TODOs are removed/cleaned up in https://github.com/kubernetes/kubernetes/pull/74966/commits/5384ce66b76541de3be156f45c25dab6e20bef03 again

I didn't want to reference commit shas in the source code TODOs as they might not be stable (-> rebases).

hoegaarden replied:
This workaround is not needed any more, we decided not to split this in multiple commits but do it all at once.

hoegaarden commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -153,24 +157,25 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
-		if unregErr := unregisterDriver(pluginName); unregErr != nil {
-			klog.Error(log("registrationHandler.RegisterPlugin failed to unregister plugin due to previous error: %v", unregErr))
+		if unregErr := p.unregisterDriver(pluginName); unregErr != nil {
+			klog.Error(log("RegisterPlugin failed to unregister plugin due to previous error: %v", unregErr))
```
I thought by using https://github.com/kubernetes/kubernetes/blob/5384ce66b76541de3be156f45c25dab6e20bef03/pkg/volume/csi/csi_util.go#L108-L110 the context might be clear enough. However you are right: more context does not hurt. Will add something back.

hoegaarden commented:
/retest

hoegaarden commented:
/retest

hoegaarden commented:
/retest

vladimirvivien commented on pkg/kubelet/kubeletfakes/fake_plugin_watcher.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+func (fake *FakePluginWatcher) AddHandlerCallCount() int {
```
Are these `xxxCallCount` and `xxxForCall` methods used anywhere, is that for testify ?

hoegaarden replied:
This is the fake, automatically generated by [counterfeiter](https://github.com/maxbrunsfeld/counterfeiter). It just generates those methods for all the methods of the interface it is meant to fake.

vladimirvivien commented:
@hoegaarden 
 I am thinking this can be made simpler still: 

1) Because CSI is a special plugin (in that it is the only way to add volume functionalities when all other plugins are deprecated), I think it's OK to augment the VolumeManager with a new method called `VolumeManager.GetCSIKubeletPluginHandler() `
That method could be added here:
https://github.com/kubernetes/kubernetes/blob/ca0c44641feadbd184da2e347d06db6b4163875d/pkg/volume/plugins.go#L445
Internally, that method can use `FindPluginByName()` then type assert to `pluginwatcher.PluginHandler`

2) This would eliminate the need to import package `csi` in the Kubelet as done here 
https://github.com/kubernetes/kubernetes/pull/74966/commits/3881f2de645cf8ec3156b33b8bd3c7a3cea0cda8#diff-bf28da68f62a8df6e99e447c4351122dR1394

3) Here, in the Kubelet, when it's time to access the CSI plugin handler, it can be done with

```go
func (kl *Kubelet) registerCSIPluginHandler() error {
	csiPluginHandler, err := kl.volumePluginMgr.GetCSIKubeletPluginHandler()
	if err != nil {
		return err
	}
    kl.pluginWatcher.AddHandler(pluginwatcherapi.CSIPlugin, csiPluginHandler)
}
```
https://github.com/kubernetes/kubernetes/pull/74966/commits/3881f2de645cf8ec3156b33b8bd3c7a3cea0cda8#diff-bf28da68f62a8df6e99e447c4351122dR1388

hoegaarden commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -197,15 +204,62 @@ func (h *RegistrationHandler) validateVersions(callerName, pluginName string, en
+
+	addr := fmt.Sprintf(csiAddrTemplate, driverName)
+	requiresV0Client := true
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
```
Currently in master there is still this feature gate in place. Should we remove the feature gate in a separate PR maybe?

https://github.com/kubernetes/kubernetes/blob/1192ca09f0a5b01771002b2773b71b8ac6745c04/pkg/volume/csi/csi_client.go#L146-L162

jsafrane replied:
Please don't remove features yet, check https://github.com/kubernetes/kubernetes/blob/da31c50da191ae1c055069b936d9e549741e3e8a/pkg/features/kube_features.go#L505 when it should be removed.

hoegaarden commented:
> Because CSI is a special plugin [...], I think it's OK to augment the VolumeManager with a new method called `VolumeManager.GetCSIKubeletPluginHandler()`

That would mean that `pkg/volume` needs to import `pkg/volume/csi`[1] which in turn imports `pkg/volume`, so we'd have an import cycle.

There might be a nice/easy way to break that cycle and I will look into that.
For what it's worth, in case that's a concern: We do not add the dependency of the kubelet on `pkg/volume/csi`, this dependency already exists today, we just change how we use it.

[1]: We need to import the `pkg/volume/csi` package to have access to the [const `CSIPluginName`](https://github.com/kubernetes/kubernetes/blob/1192ca09f0a5b01771002b2773b71b8ac6745c04/pkg/volume/csi/csi_plugin.go#L50) and to have access to the [type `csiPlugin`](https://github.com/kubernetes/kubernetes/blob/1192ca09f0a5b01771002b2773b71b8ac6745c04/pkg/volume/csi/csi_plugin.go#L68) to do the type assertion from `volume.VolumePlugin`.

vladimirvivien commented:
@hoegaarden 
Thanks for taking a look. I don't think there would by cyclic issues.
- pkg/volume would not need to import `pkg/volume/csi`
- `pkg/volume` would need to import `pluginwatcher.PluginHandler`
- method `VolumeManager.GetCSIPluginHandler` would return a `pluginwatcher.PluginHander` not CSIPlugin.

vladimirvivien commented:
@hoegaarden 
After some thought experiment, I see that there will in fact be an import of csi package as you pointed out.  But this can be fixed (i think) by creating a new `VolumePlugin` type:

- Add a new volume type in pkg/volume/plugins.go
```go
type interface KubeletWatchableVolumePlugin {
    VolumePlugin
    GetWatcherHandler() pluginswatcher.PluginHandler
}
```

- Add method `VolumePluginMgr`
```go
func (pm *VolumePluginMgr) GetCSIKubeletPluginHandler() pluginswatcher.PluginHandler {
    volumePlugin, _ := pm.FindPluginByName(name)
    return volumePlugin.(pluginswatcher.PluginHandler)
}

- Then in kubelet.go
```go
func (kl *Kubelet) registerCSIPluginHandler() error {
	csiPluginHandler, err := kl.volumePluginMgr.GetCSIKubeletPluginHandler()
	if err != nil {
		return err
	}
    kl.pluginWatcher.AddHandler(pluginwatcherapi.CSIPlugin, csiPluginHandler)
}
```

hoegaarden commented:
/retest

hoegaarden commented:
/retest

hoegaarden commented:
@vladimirvivien 

_**PTAL**_

I implemented the approach you [layed out](https://github.com/kubernetes/kubernetes/pull/74966#issuecomment-484541193).

As discussed
- I squashed all the commits
- the packages [`k8s.io/kubernetes/pkg/volume/csi/identity`](https://github.com/kubernetes/kubernetes/pull/74966/files#diff-2ac890d0b483350ea5112bbf9367b62d) feels a bit strange, but for now the fastest way to break the import cycle

hoegaarden commented:
/retest

vladimirvivien commented on pkg/volume/csi/identity/identity.go:
```diff
@@ -0,0 +1,22 @@
+limitations under the License.
+*/
+
+package identity
```
Please comment why this is needed and how it is used in the code.

vladimirvivien commented on pkg/volume/plugins.go:
```diff
@@ -312,6 +314,17 @@ type BlockVolumePlugin interface {
 }
 
+// KubeletWatchableVolumePlugin is an extended interface of VolumePlugin and/
+// describes plugins that can be used as plugin handlers.
```
Please make it clear this is used by the Kubelet Plugin watcher for discovery.

vladimirvivien commented on pkg/volume/plugins.go:
```diff
@@ -1028,6 +1041,22 @@ func (pm *VolumePluginMgr) FindNodeExpandablePluginByName(name string) (NodeExpa
+
+	kubeletWatchablePlugin, ok := volumePlugin.(KubeletWatchableVolumePlugin)
+	if !ok {
+		// TODO (hoegaarden) Can this even happen?
```
Yes, this can happen in a case where code changes and no longer satisfies the interface.

hoegaarden replied:
I think, that cannot really happen, because we have the compile time assertion here: https://github.com/kubernetes/kubernetes/pull/74966/files/7b30c10ec5ba1578c5306df33722c9fb188d7f31#diff-f52821ec6236e86b02ca334533395c91R1015

But someone could remove that somewhen, so we def. need to check the type here.

In any case, I will remove the comment.

vladimirvivien commented on pkg/kubelet/kubelet.go:
```diff
@@ -1215,6 +1214,13 @@ type Kubelet struct {
 }
 
+//go:generate counterfeiter . pluginWatcher
+type pluginWatcher interface {
```
I would comment why this type is introduced (i.e. needed to make testing easier, etc).

vladimirvivien reviewed (commented):
I think this is great.
One suggestion, edit the PR and give a summary of the design and why things are move so drastically around.  This will help reviewers (who dont have context) to quickly see what is going on.  For instance, shifting the responsibility of the types around to remove the package level singleton as a way for kubelet to reach CSI plugin architecture.

hoegaarden commented:
/test pull-kubernetes-e2e-gce-100-performance

vladimirvivien commented:
@jsafrane  @msau42 PTAL at this refactoring PR for second opinions. LGTM.

vladimirvivien commented:
@kubernetes/sig-node-pr-reviews PTAL

jsafrane commented on pkg/kubelet/kubeletfakes/fake_plugin_watcher.go:
```diff
@@ -0,0 +1,186 @@
+/*
+Copyright The Kubernetes Authors.
```
year is missing?

hoegaarden replied:
This is a generated file, thus uses the [boilerplate.generatego.txt](https://github.com/kubernetes/kubernetes/blob/f780ac028b444ed15d03de755c12fcf84c783286/hack/boilerplate/boilerplate.generatego.txt) header, without the year.

jsafrane commented on pkg/volume/csi/constants/constants.go:
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Wrong year

hoegaarden replied:
Changed. Thanks for the review, @jsafrane!

jsafrane reviewed (commented):
lgtm-ish with minor nits

/approve
the volume plugin changes.

hoegaarden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hoegaarden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hoegaarden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hoegaarden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

timmycarr commented:
Hi! Friendly reminder from your release team: we are starting the code freeze for 1.15 tomorrow EOD. Just checking in to see if this is still planned for the 1.15 cycle?

soggiest commented:
/milestone v1.16

hoegaarden commented:
/test pull-kubernetes-integration

hoegaarden commented:
/test pull-kubernetes-integration

hoegaarden commented:
@vladimirvivien @cheftako @sjpotter @msau42 @jsafrane
The PR has been rebased and all tests are green currently. Could you please review (again) and let me know what you think? Would be great if that went in early in the 1.16 cycle.

hoegaarden commented:
/test pull-kubernetes-e2e-gce

msau42 commented:
/assign @jsafrane

hoegaarden commented:
/retest

Pothulapati commented:
Hello! I'm part of the bug triage team for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th, which means that there should be a PR ready (and merged) until then.

Do we still target this for 1.16? If not please re-tag the issue to the planned milestone.

hoegaarden commented:
/milestone v1.17

hoegaarden commented:
/retest

ttousai commented:
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on Nov. 18. Is this PR still intended for 1.17?

jsafrane commented:
Sorry, I missed this one.
/lgtm
/approve

hoegaarden commented:
/assign @derekwaynecarr

for the changes to
- `pkg/kubelet/...`
  - change the registration mechanism of the CSI plugin handler
  - adding the pluginmanager fakes and the `go generate` stanza to re-generate the fakes
- `pkg/controller/...`
  - import restrictions

josiahbjorgaard commented:
@hoegaarden Is this PR still intended to be merged before the v1.17 code freeze? It is this Thursday, Nov. 14 at 5PM. Otherwise, bump to v1.18?

jsafrane commented:
we'd like to catch 1.17

/assign @derekwaynecarr @tallclair 
for approval.

ttousai commented:
@tallclair @derekwaynecarr code freeze is tomorrow, Thursday 14th. Please review and approve this PR before then.

hoegaarden commented:
/test pull-kubernetes-node-e2e-containerd

hoegaarden commented:
The 'pull-kubernetes-node-e2e-containerd' failure is probably unrelated to the changes in this PR, see https://github.com/kubernetes/kubernetes/issues/85040 .

liggitt commented:
if this is cleanup only, I'd recommend deferring it to 1.18

liggitt commented:
/milestone clear

hoegaarden commented:
/milestone v1.18

hoegaarden commented:
@jsafrane -- I just rebased again, maybe we can get it into `v1.18`?

Also, as already mentioned, the failing test `pull-kubernetes-node-e2e-containerd` does not seem to be related to my change (see: https://github.com/kubernetes/kubernetes/issues/85040)

jsafrane commented:
/lgtm

hoegaarden commented:
/retest

hoegaarden commented:
/remove-sig cli
/remove-area kubectl
... as those just creeped in via a faulty rebase (https://github.com/kubernetes/kubernetes/pull/74966#event-2924599092).

hoegaarden commented:
@tallclair @derekwaynecarr
Any chance you can review that?

jsafrane commented:
/lgtm

tallclair commented on pkg/kubelet/kubelet.go:
```diff
@@ -1385,15 +1383,30 @@ func (kl *Kubelet) initializeRuntimeDependentModules() {
 	// Adding Registration Callback function for CSI Driver
-	kl.pluginManager.AddHandler(pluginwatcherapi.CSIPlugin, plugincache.PluginHandler(csi.PluginHandler))
+	if err := kl.registerCSIPluginHandler(); err != nil {
+		kl.recorder.Eventf(kl.nodeRef, v1.EventTypeWarning, events.KubeletSetupFailed, err.Error())
```
Is this a synchronous call, or is it possible that for the following fatal call to happen before the event is sent to the apiserver?

tallclair commented on pkg/kubelet/kubelet.go:
```diff
@@ -1385,15 +1383,30 @@ func (kl *Kubelet) initializeRuntimeDependentModules() {
-	kl.pluginManager.AddHandler(pluginwatcherapi.CSIPlugin, plugincache.PluginHandler(csi.PluginHandler))
+	if err := kl.registerCSIPluginHandler(); err != nil {
+		kl.recorder.Eventf(kl.nodeRef, v1.EventTypeWarning, events.KubeletSetupFailed, err.Error())
+		klog.Fatalf("failed to register CSI Plugin Handler. err: %v", err)
```
Under what circumstances would an error occur here? Adding a `Fatal` call in the kubelet startup path makes me nervous.

derekwaynecarr commented:
defer to @tallclair for final review

jtslear commented:
Hello, @hoegaarden, @cheftako, and @sjpotter 
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this PR still intended for milestone 1.18?  It looks like we are awaiting reviews and there are some conflicting files.

jtslear commented:
Hello, @hoegaarden, @tallclair 
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this PR still intended for milestone 1.18? There hasn't been any movement on this since the end of January.  We are pending reviews, and conflicts.

cc @kubernetes/sig-apps-pr-reviews, @kubernetes/sig-node-pr-reviews, @kubernetes/sig-storage-pr-reviews

jtslear commented:
With no response during the course of this release, I'm going to clear the milestone tracking of this PR.

/milestone clear

cheftako commented:
Are we still trying to merge this or should we just close it?

hoegaarden commented:
I think I have more bandwidth again to look into that. I have no idea what happened in the CSI world in the last months, so I will check if I can rebase it or close it otherwise.

/remove-lifecycle stale

hoegaarden commented:
/retest
